app/controllers/nonsipp/ReturnSubmittedController.scala

```diff
@@ -0,0 +1,155 @@
+import models.SchemeId.Srn
+import models.requests.DataRequest
+import models.{DateRange, Mode}
+import pages.nonsipp.accountingperiod.AccountingPeriods
+import pages.nonsipp.{ReturnSubmittedPage, WhichTaxYearPage}
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.libs.json.Writes._
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents, Result}
+import services.{SaveService, SchemeDateService}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import viewmodels.DisplayMessage
+import viewmodels.DisplayMessage._
+import viewmodels.implicits._
+import viewmodels.models.SubmissionViewModel
+import views.html.SubmissionView
+
+import java.time.LocalDateTime
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class ReturnSubmittedController @Inject()(
+  override val messagesApi: MessagesApi,
+  identifyAndRequireData: IdentifyAndRequireData,
+  saveService: SaveService,
+  view: SubmissionView,
+  dateService: SchemeDateService,
+  config: FrontendAppConfig,
+  val controllerComponents: MessagesControllerComponents
+)(implicit ec: ExecutionContext)
+    extends FrontendBaseController
+    with I18nSupport {
+
+  def onPageLoad(srn: Srn, mode: Mode): Action[AnyContent] = identifyAndRequireData(srn).async { implicit request =>
+    getReturnPeriods(srn) { returnPeriods =>
+      getOrSaveSubmissionDate(srn).map { submissionDate =>
+        Ok(
+          view(
+            viewModel(
+              request.schemeDetails.schemeName,
+              request.minimalDetails.email,
+              returnPeriods,
+              submissionDate,
+              config.urls.pensionSchemeEnquiry,
+              config.urls.managePensionsSchemes.dashboard
+            )
+          )
+        )
+      }
+    }
+  }
+
+  private def getOrSaveSubmissionDate(srn: Srn)(implicit request: DataRequest[_]): Future[LocalDateTime] =
+    request.userAnswers.get(ReturnSubmittedPage(srn)) match {
+      case Some(submissionDate) => Future.successful(submissionDate)
+      case None =>
+        val submissionDate = dateService.now()
+        for {
+          updatedUserAnswers <- Future.fromTry(request.userAnswers.set(ReturnSubmittedPage(srn), submissionDate))
+          _ <- saveService.save(updatedUserAnswers)
+        } yield submissionDate
+    }
+
+  private def getReturnPeriods(
+    srn: Srn
+  )(f: NonEmptyList[DateRange] => Future[Result])(implicit request: DataRequest[_]): Future[Result] = {
+    val accountingPeriods = request.userAnswers.list(AccountingPeriods(srn))
+
+    if (accountingPeriods.isEmpty) {
+      request.userAnswers.get(WhichTaxYearPage(srn)) match {
+        case Some(returnPeriod) => f(NonEmptyList.of(returnPeriod))
+        case None => Future.successful(Redirect(controllers.routes.JourneyRecoveryController.onPageLoad()))
+      }
+    } else {
+      f(NonEmptyList.fromListUnsafe(accountingPeriods)) // safe as empty check is performed
```

Comment:
Instead of using `fromListUnsafe` why not use `fromList` and use `map` and `getOrElse` on the result?

Something like
```
NonEmptyList.fromList(accountingPeriods)
  .map(f)
  .getOrElse {
    request.userAnswers.get(WhichTaxYearPage(srn)) match {
        case Some(returnPeriod) => f(NonEmptyList.of(returnPeriod))
        case None => Future.successful(Redirect(controllers.routes.JourneyRecoveryController.onPageLoad()))
      }
  }
```

Reply:
the getOrElse is functional unreachable but I suppose your proposal is easier to understand over an unsafe method